File: src/funcs.rs
use std::mem;
use std::env;
use std::ffi::OsStr;
use std::io::{Read, BufRead, BufReader};

use sha2::{Sha256, Digest};

use crate::types::{BlobId, BlobIdStr, BlobIdRef, BlobRef, Status, Result};
use crate::{log_err, log_ret_err, log_err_ret_val};


const CHUNK_SIZE: usize = 64 * 1024;


pub fn env(name_in: impl AsRef<OsStr>) -> Result<Box<str>> {
    let name = name_in.as_ref();
    match env::var(name) {
        Ok(val) => Ok(val.into()),
        Err(_error) => {
            let name_str = name.to_string_lossy();
            eprintln!("Missing required environment variable: {name_str}");
            Err(Status::InternalError)
        },
    }
}

pub fn blob_hash(blob: BlobRef) -> BlobId {
    BlobId(Sha256::digest(blob).into())
}

pub fn hash_copy(input: &mut impl Read, output: &mut [u8])
        -> Result<BlobId> {
    let mut hasher = Sha256::new();
    for chunk in output.chunks_mut(CHUNK_SIZE) {
        log_ret_err!(input.read_exact(chunk));
        hasher.update(chunk);
    }
    Ok(BlobId(hasher.finalize().into()))
}

pub fn hash(input: &mut impl Read) -> Result<BlobId> {
    let mut reader = BufReader::with_capacity(CHUNK_SIZE, input);
    let mut hasher = Sha256::new();
    while log_ret_err!(reader.has_data_left()) {
        let chunk = log_ret_err!(reader.fill_buf());
        hasher.update(chunk);
        let chunk_size = chunk.len();
        reader.consume(chunk_size);
    }
    Ok(BlobId(hasher.finalize().into()))
}

pub const fn code8(bytes: & [u8; 8]) -> u64 {
    u64::from_le_bytes(*bytes)
}

pub fn hash_to_str(blob_id: BlobIdRef) -> BlobIdStr {
    fn byte_to_hex(byte: u8) -> [u8; 2] {
        [nibble_to_hex_digit(byte >> 4), nibble_to_hex_digit(byte & 0xf)]
    }
    fn nibble_to_hex_digit(nibble: u8) -> u8 {
        match nibble {
            0..10 => nibble + b'0',
            10..16 => nibble - 10 + b'a',
            _ => panic!("nibble_to_hex_digit: invalid nibble: {nibble}"),
        }
    }

    let BlobId(hash) = blob_id;
    unsafe {mem::transmute(hash.map(byte_to_hex))}
}

pub fn str_to_hash(string: &[u8]) -> Result<BlobId> {
    fn hex_to_byte([hi, lo]: [u8; 2]) -> Result<u8> {
        let byte = hex_digit_to_nibble(hi)? << 4 | hex_digit_to_nibble(lo)?;
        Ok(byte)
    }
    fn hex_digit_to_nibble(digit: u8) -> Result<u8> {
        match digit {
            b'0'..=b'9' => Ok(digit - b'0'),
            b'a'..=b'f' => Ok(digit - b'a' + 10),
            _ => {
                eprintln!("hex_digit_to_nibble: invalid hex digit '{digit}'");
                Err(Status::InternalError)
            },
        }
    }
    let hash_str = log_ret_err!(BlobIdStr::try_from(string));
    let hash_str_chunks: [[u8; 2]; _] = unsafe {mem::transmute(hash_str)};
    let hash = hash_str_chunks.try_map(hex_to_byte)?;
    Ok(BlobId(hash))
}

File: zig/funcs.zig
const std = @import("std");
const Environ = std.process.Environ;
const Hasher = std.crypto.hash.sha2.Sha256;

const ty = @import("types.zig");
const mem = @import("mem.zig");

const CHUNK_SIZE: usize = 64 * 1024;

pub fn getEnv(env: *Environ.Map, name: []const u8) ty.Err![]const u8 {
    return env.get(name) orelse out: {
        debug("Missing required environment variable: {s}\n", .{name});
        break :out ty.Err.Internal;
    };
}

pub fn debug(comptime format: []const u8, args: anytype) void {
    std.debug.print(format, args);
}

pub fn hashBlob(blob: *ty.Blob) !ty.BlobId {
    return switch (blob.*) {
        .stream => |*in| out: {
            var hasher = Hasher.init(.{});
            var chunk = try mem.alloc(u8, CHUNK_SIZE);
            defer mem.free(chunk);
            while (in.bytes_left > 0) {
                const slice = chunk[0 .. @min(in.bytes_left, CHUNK_SIZE)];
                try in.reader.readSliceAll(slice);
                hasher.update(slice);
            }
            break :out hasher.finalResult();
        },
        .file => |*file| out: {
            const mmap_opts: std.Io.File.MemoryMap.CreateOptions = .{
                .len = try file.file.length(file.io),
            };
            var mem_map = try file.file.createMemoryMap(file.io, mmap_opts);
            defer mem_map.destroy(file.io);
            break :out hashMemory(mem_map.memory);
        },
        .memory => |bytes| hashMemory(bytes),
    };
}

pub fn hashCopyBlob(input: *std.Io.Reader, output: []u8) !ty.BlobId {
    var hasher = Hasher.init(.{});
    var index: usize = 0;
    while (index < output.len) : (index += CHUNK_SIZE) {
        const chunk = output[index .. @min(index + CHUNK_SIZE, output.len)];
        try input.readSliceAll(chunk);
        hasher.update(chunk);
    }
    return hasher.finalResult();
}

pub fn hashMemory(blob: []u8) ty.BlobId {
    var blob_id: ty.BlobId = undefined;
    Hasher.hash(blob, &blob_id, .{});
    return blob_id;
}

pub fn hashBytesToHex(hash: ty.BlobId) ty.BlobIdStr {
    var string: ty.BlobIdStr = undefined;
    for (hash, 0..) |byte, index| {
        @memcpy(sliceChunk(2, @as([]u8, &string), index), &byteToHex(byte));
    }
    return string;
}

fn byteToHex(byte: u8) [2]u8 {
    const pair: NibblePair = @bitCast(byte);
    return .{ nibbleToHexDigit(pair.hi), nibbleToHexDigit(pair.lo) };
}

fn nibbleToHexDigit(nibble: u4) u8 {
    return switch (nibble) {
        0x0...0x9 => @as(u8, nibble) + '0',
        0xa...0xf => @as(u8, nibble) - 0xa + 'a',
    };
}

pub fn hashHexToBytes(string: []const u8) ty.Err!ty.BlobId {
    if (string.len != 64) {
        debug("hashHexToBytes: invalid length string: '{s}'\n", .{string});
        return ty.Err.Internal;
    }
    var hash: ty.BlobId = undefined;
    for (&hash, 0..) |*byte, index| {
        const slice = sliceChunk(2, string, index);
        byte.* = try hexToByte(@ptrCast(slice));
    }
    return hash;
}

fn hexToByte(hex_pair: *const[2]u8) ty.Err!u8 {
    const hi, const lo = hex_pair.*;
    const nibbles: NibblePair = .{.hi = try hexDigitToNibble(hi),
                                  .lo = try hexDigitToNibble(lo)};
    return @bitCast(nibbles);
}

fn hexDigitToNibble(digit: u8) ty.Err!u4 {
    return switch (digit) {
        '0'...'9' => @intCast(digit - '0'),
        'a'...'f' => @intCast(digit - 'a' + 0xa),
        else => out: {
            debug("hexDigitToNibble: " ++
                  "invalid hex digit: '{c}'\n", .{digit});
            break :out ty.Err.Internal;
        },
    };
}

const NibblePair = packed struct(u8) {
    lo: u4,
    hi: u4,
};

fn sliceChunk(comptime chunk_size: usize, slice: anytype, chunk_num: usize)
        @TypeOf(slice) {
    return slice[chunk_size * chunk_num .. chunk_size * (chunk_num + 1)];
}

pub fn pairGen(A: type, B: type) type {
    const PairStruct = packed struct {
        a: A,
        b: B,
    };
    const Pair = struct {
        pub fn make(a: A, b: B) PairStruct {
            return .{.a = a, .b = b};
        }
    };
    return Pair;
}
